### PR TITLE
Specify details of MP_RETIRE_CONNECTION_ID

### PR DESCRIPTION
**Pull request by huitema**
Addresses issue #368.

### CODE REVIEW

File: draft-ietf-quic-multipath.md [hunk 1]
```diff
@@ -1197,12 +1197,22 @@ Path Identifier:
 Note that, other than for the NEW_CONNECTION_ID frame of {{Section 19.15 of QUIC-TRANSPORT}},
 the sequence number applies on a per-path context.
 This means different connection IDs on different paths may have the same
-sequence number value. Respectively, the Retire Prior To field indicates which connection IDs
-should be retired for the path with the Path ID in the Path Identifier field.
+sequence number value. However, the value of the Connection ID MUST be unique
+across all paths.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1201]: I don't think we need normative language here as this is already specified in RFC9000. Let's not specify this again.

Also note that 3.5 already says this:

```
As further specified in [Section 5.1](https://rfc-editor.org/rfc/rfc9000#section-5.1) of [[QUIC-TRANSPORT](https://quicwg.org/multipath/draft-ietf-quic-multipath.html#QUIC-TRANSPORT)] connection IDs cannot be issued more than once on the same connection and therefore are unique for the scope of the connection, regardless of the associated Path ID.
```

We can repeat it here (but not normatively).

File: draft-ietf-quic-multipath.md [hunk 2]
```diff
@@ -1197,12 +1197,22 @@ Path Identifier:
...
+
+The Retire Prior To field indicates which connection IDs
+should be retired among those that share the Path ID in the Path Identifier field.
+Connection IDs associated with different path IDs are not affected.
 
 Note that the NEW_CONNECTION_ID frame can only be used to issue or retire
 connection IDs for the initial path with Path ID 0.
 
+Endpoints SHOULD NOT issue updates of the Retire Prior To field for a specific path ID
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1210]: Let's not use normative language here is it is already specified normatively in RFC9000.

File: draft-ietf-quic-multipath.md [hunk 3]
```diff
@@ -1238,8 +1248,31 @@ Note that the RETIRE_CONNECTION_ID frame can only be used to retire
...
+in which the frame is contained. The peer MAY treat this as a connection
+error of type PROTOCOL_VIOLATION. If the multipath extension is negotiated,
+the same error is detected if a RETIRE_CONNECTION_ID frame is received,
+the path ID of the Destination Connection ID field of the packet is 0,
+and the sequence number of that connection ID matches Sequence Number
+field of the RETIRE_CONNECTION_ID frame.
+
+These last two paragraphs change on
+the processing of RETIRE_CONNECTION_ID frames specified in
+{{Section 19.17 of QUIC-TRANSPORT}}.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1275]: I don't think we should repeat all normative language from RFC9000 here. I would rather prefer something like:

```
The processing as described In {{Section 19.17 of QUIC-TRANSPORT}} applies for MP_RETIRE_CONNECTION_ID frames per path, while the processing of an RETIRE_CONNECTION_ID frame is only applied for Path ID 0.
```

**Review by mirjak — CHANGES_REQUESTED** [review]
I don't think we should repeat normative requirements from RFC9000 here.

File: draft-ietf-quic-multipath.md (hunk 3 above)
Comment (huitema) [reply draft-ietf-quic-multipath.md:1275]: What changes is the processing of the plain RETIRE_CONNECTION_ID, and that is a deviation from RFC 9000. Implementations need not just implement the new MP_RETIRE_CONNECTION_ID, they need to change the way they implemented RETIRE_CONNECTION_ID.

File: draft-ietf-quic-multipath.md (hunk 2 above)
Comment (mirjak) [reply draft-ietf-quic-multipath.md:1210]: We can have more than one sentence but I don't think we should normatively rephrase things from RFC9000. So if you add text that ensure it is well understood what we mean without using normative language that would be my preferred option.

File: draft-ietf-quic-multipath.md (hunk 3 above)
Comment (mirjak) [reply draft-ietf-quic-multipath.md:1275]: Yes, I thought that exactly what my proposed sentence point out explicitly. I believe saying this is actually more clear than repeating the text from RFC9000.

File: draft-ietf-quic-multipath.md (hunk 2 above)
Comment (huitema) [reply draft-ietf-quic-multipath.md]: OK, updating the text.

File: draft-ietf-quic-multipath.md (hunk 3 above)
Comment (huitema) [reply draft-ietf-quic-multipath.md]: OK, updating the text.

File: draft-ietf-quic-multipath.md (hunk 1 above)
Comment (huitema) [reply draft-ietf-quic-multipath.md]: OK, removed the last sentence of that paragraph.

**Review by Yanmei-Liu — APPROVED** [review]
Looks good!

File: draft-ietf-quic-multipath.md [hunk 4]
```diff
@@ -1197,12 +1197,22 @@ Path Identifier:
...
+sequence number value.
+
+The Retire Prior To field indicates which connection IDs
+should be retired among those that share the Path ID in the Path Identifier field.
+Connection IDs associated with different path IDs are not affected.
 
 Note that the NEW_CONNECTION_ID frame can only be used to issue or retire
 connection IDs for the initial path with Path ID 0.
 
+The last paragraph of {{Section 5.1.2 of QUIC-TRANSPORT}} specified how to
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1209]:
```suggestion
The last paragraph of {{Section 5.1.2 of QUIC-TRANSPORT}} specifies how to
```

File: draft-ietf-quic-multipath.md [hunk 5]
```diff
@@ -1238,8 +1248,14 @@ Note that the RETIRE_CONNECTION_ID frame can only be used to retire
 connection IDs for the initial path with Path ID 0.
 
 As the MP_NEW_CONNECTION_ID frames applies the sequence number per path,
-the sequence number in the MP_RETIRE_CONNECTION_ID frame
-also needs to be considered in the context of the Path Identifier field.
+the sequence number in the MP_RETIRE_CONNECTION_ID frame is also per
+path. The MP_RETIRE_CONNECTION_ID frame retires the Connection ID with
+the specified path ID and sequence number.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1253]:
```suggestion
the specified Path ID and sequence number.
```

File: draft-ietf-quic-multipath.md [hunk 6]
```diff
@@ -1197,12 +1197,22 @@ Path Identifier:
...
+Connection IDs associated with different path IDs are not affected.
 
 Note that the NEW_CONNECTION_ID frame can only be used to issue or retire
 connection IDs for the initial path with Path ID 0.
 
+The last paragraph of {{Section 5.1.2 of QUIC-TRANSPORT}} specifies how to
+verify the Retire Prior To field of an incoming NEW_CONNECTION_ID frame.
+The same rule
+applies for MP_RETIRE_CONNECTION_ID frames, but it applies per path. After the
+multipath extension is isolated, the rule
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:1213]:
```suggestion
multipath extension is negotiated successfully, the rule
```

**Review by mirjak — APPROVED** [review]
This is a clarification and I think we can merge it.

**Pull request merged**